### ISSUE TITLE
About the implementation of mask average pool.

### ISSUE THREAD

**Issue by lartpang** [chenqi1126/SIPE#5]
https://github.com/chenqi1126/SIPE/blob/e0c6a3bc578a6bade2cf85fe34a1793dce71170d/network/resnet50_SIPE.py#L58

![image](https://user-images.githubusercontent.com/26847524/187104863-b75ce770-f0b6-4953-9e15-80d6ff82a7bb.png)

The denominator of the two is not the same. The code is the area of the entire feature map (h*w), while the formula is the area of the foreground.

**Comment by IMSEMZPZ** [comment]
您好，我是Pangze Cheung。我已经收到您的邮件，我会尽快回复您的！谢谢！-------------------------------------------------------------------------------------------------------------Best Wishes For YouPangze Cheung

**Comment by chenqi1126** [comment]
Hi pang, the 'crop_feature' only contains the features of the foreground since the unrelated parts have been eliminated by seeds (see Line 56-57).

**Comment by lartpang** [comment]
@chenqi1126 

Thank you for your reply. 

https://github.com/chenqi1126/SIPE/blob/e0c6a3bc578a6bade2cf85fe34a1793dce71170d/network/resnet50_SIPE.py#L56-L58

Although the background is eliminated by seed here, this only corresponds to the calculation process of the numerator in the formula.
But the denominator is calculated differently than the formula.
According to the meaning of the formula, **mask average pooling only calculates the average value on a specific area (R^i_k=1)**, while in the current implementation, it is calculated on the entire feature, **which leads to a larger denominator**.

**Comment by chenqi1126** [comment]
@lartpang 

Thanks for your attention. 

Yes - It is a little different between formula and implementation. As you mentioned, the implementation would lead to a larger denominator and produce a scaled value compared to the formula.  But this scaled value does not affect the final result since it is normalized in the following similarity calculation. (see below, the official PyTorch document is [here](https://pytorch.org/docs/stable/generated/torch.nn.functional.cosine_similarity.html).)
Current implementation simplifies the calculation and avoids division by zero caused by unstable seeds in the early training stage.
https://github.com/chenqi1126/SIPE/blob/e0c6a3bc578a6bade2cf85fe34a1793dce71170d/network/resnet50_SIPE.py#L62

**Comment by lartpang** [comment]
Ok, got it, thanks!
